AnyText/AnyText.Lsp/LspServer.SemanticToken.cs

```diff
@@ -0,0 +1,221 @@
+using LspTypes;
+using Newtonsoft.Json.Linq;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace NMF.AnyText
+{
+    public partial class LspServer
+    {
+        private readonly Dictionary<string, uint[]> _previousSemanticTokens = new();
+
+        /// <summary>
+        ///     Handles the <c>textDocument/semanticTokens/full</c> request from the client. This is used to retrieve all semantic
+        ///     tokens for a document.
+        /// </summary>
+        /// <param name="arg">The JSON token containing the parameters of the request. (SemanticTokensParams)</param>
+        /// <returns>A <see cref="SemanticTokens" /> object containing the full set of semantic tokens for the document.</returns>
+        public SemanticTokens QuerySemanticTokens(JToken arg)
+        {
+            var semanticTokensParams = arg.ToObject<SemanticTokensParams>();
+            var uri = semanticTokensParams.TextDocument.Uri;
+            
+            if (!_documents.TryGetValue(uri, out var document))
+                return new SemanticTokens { ResultId = null, Data = Array.Empty<uint>() };
+
+            var semanticElements = document.GetSemanticElementsFromRoot().ToArray();
+
+            var resultId = Guid.NewGuid().ToString();
+            _previousSemanticTokens[resultId] = semanticElements;
+
+            return new SemanticTokens
+            {
+                ResultId = resultId,
+                Data = semanticElements
+            };
+        }
+
+        /// <summary>
+        ///     Handles the <c>textDocument/semanticTokens/full/delta</c> request from the client. This is used to retrieve only
+        ///     the changes (delta) in semantic tokens for a document.
+        /// </summary>
+        /// <param name="arg">The JSON token containing the parameters of the request. (SemanticTokensDeltaParams)</param>
+        /// <returns>A <see cref="SemanticTokensDelta" /> object containing only the delta of semantic tokens for the document.</returns>
+        public SemanticTokensDelta QuerySemanticTokensDelta(JToken arg)
```

Comment:
Der Ansatz ist ja jetzt, sich im Server beide anzuschauen und zu schauen, wo da jetzt der Unterschied liegt. Das ist eigentlich mehr Aufwand, als jedes mal einfach die Token komplett neu zu generieren, weil letzteres ja ein Teilschritt davon ist. Daher sehe ich hier zwei Optionen:

1. Wir behalten den Code, sagen aber in den Capabilities, dass wir kein Delta können und zwingen den Client damit, immer alle Tokens neu anzufordern, weil das mehr Aufwand bedeutet.
2. Wir fügen noch eine Logik ein, die sich die TextEdits anschaut und sich dann vormerkt, ab welchem Index die Token neu berechnet werden müssen.

Ehrlich gesagt tendiere ich zu Option 1, da Option 2 wahrscheinlich nicht einfach werden würde, ich mir auch noch nicht sicher bin, wie viel Effizienz das Ganze bringt.

Reply:
Aktuell ist Option 1 umgesetzt.